sys/mips/pmap.c

```diff
@@ -552,33 +548,27 @@ void pmap_delete(pmap_t *pmap) {
   pool_free(P_PMAP, pmap);
 }
 
-#define L1_SPACE_SIZE (PAGESIZE * PAGESIZE / sizeof(pte_t))
-
 /*
  * Increase usable kernel virtual address space to at least maxkvaddr.
  * Allocate page table (level 1) if needed.
  */
-vaddr_t pmap_growkernel(vaddr_t maxkvaddr) {
-  assert(maxkvaddr > atomic_load(&pmap_maxkvaddr));
+void pmap_growkernel(vaddr_t maxkvaddr) {
+  assert(mtx_owned(&vm_kernel_end_lock));
+  assert(maxkvaddr > vm_kernel_end);
 
   pmap_t *pmap = pmap_kernel();
   vaddr_t va;
 
-  /*
-   * For 8 pages used by kernel we need 1 page for KASAN.
-   * shadow map is grown by a multiple of L1_SPACE_SIZE.
-   */
-  maxkvaddr = roundup(maxkvaddr, 8 * L1_SPACE_SIZE);
+  maxkvaddr = roundup(maxkvaddr, PAGESIZE);
 
   WITH_MTX_LOCK (&pmap->mtx) {
-    for (va = pmap_maxkvaddr; va <= maxkvaddr; va += L1_SPACE_SIZE) {
+    for (va = vm_kernel_end; va < maxkvaddr; va += PAGESIZE) {
```

Comment:
Isn't incrementing `va` by `PAGESIZE` a bit inefficient? The step size used to be `L1_SPACE_SIZE` here. Same for AArch64.

Reply:
For `PAGESIZE` `pmap_growkernel` is called three times.
For `L1_SPACE_SIZE` two times.

Reply:
I'm not concerned about how many times `pmap_growkernel()` is called, I'm concerned about how many times `pmap_add_pde()` is called. For instance, if the current PTE index of `vm_kernel_end` is 0 (i.e. it's aligned to `L1_SPACE_SIZE`), and we call `pmap_growkernel(vm_kernel_end + L1_SPACE_SIZE)`, Then we will call `pmap_add_pde()` `L1_SPACE_SIZE / PAGESIZE` times, even though calling it once would suffice. Is my thinking correct?

Reply:
Oh, now I see what you mean.
We will call `pmap_add_pde()` exactly the same number of times in both cases.
But now we will walk through page table `L1_SPACE_SIZE / PAGESIZE` times in `PDE_OF` - similarly for AArch64.
Do you want to revert that change and go back to align `maxkvaddr` to `L1_SPACE_SIZE`?

Reply:
Yes, that's what I meant - unnecessary page table walks.
I think aligning `maxkvaddr` to `L1_SPACE_SIZE` and using steps of `L1_SPACE_SIZE` is the best way to solve this.

Reply:
done